### PR TITLE
[chore] Build and publish application hosted in the repository for the Golang instrumentation E2E test

### PR DESCRIPTION
**Pull request by iblancasa**
Related to https://github.com/open-telemetry/opentelemetry-operator/issues/2073.

The test migration will be done in a follow up PR after publishing the image.

### CODE REVIEW

File: tests/instrumentation-e2e-apps/golang/main.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+func main() {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Println("Hi")
+	})
+
+	server := &http.Server{
+		Addr:              ":8080",
+		ReadHeaderTimeout: 3 * time.Second,
```
Comment (swiatekm-sumo) [inline tests/instrumentation-e2e-apps/golang/main.go:30]: Make these constants for clarity?

**Review by swiatekm-sumo — APPROVED** [review]
Are we planning to make the publishing workflow reusable? Seems like it's almost identical for each image.

**Comment by jaronoff97** [comment]
+1 to the above, i assumed we would re-use a single workflow to publish all the images instead of one for each

**Comment by pavolloffay** [comment]
> re-use a single workflow to publish all the images instead of one for each

+1

File: tests/instrumentation-e2e-apps/golang/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FROM golang:1.21.1 as builder
```
Comment (pavolloffay) [inline tests/instrumentation-e2e-apps/golang/Dockerfile:1]: Any objections on using the apline version? We use it for the operator image

```
golang                                                                                    1.21.1              57ca605b665e   12 days ago     814MB
golang                                                                                    1.21-alpine         09df25511440   13 days ago     222MB
```

File: .github/workflows/publish-autoinstrumentation-e2e-images.yaml [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    runs-on: ubuntu-20.04
+
+    steps:
+      - uses: actions/checkout@v3
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ghcr.io/open-telemetry/opentelemetry-operator/golang-test
```
Comment (pavolloffay) [inline .github/workflows/publish-autoinstrumentation-e2e-images.yaml:31]: nit: `/e2e-test-app-golang`

**Comment by iblancasa** [comment]
> Are we planning to make the publishing workflow reusable? Seems like it's almost identical for each image.

Yep. But we need to merge something to start doing them reusable or something.

**Review by pavolloffay — APPROVED** [review]

**Pull request merged**